src/extra/components/adq.py

```diff
@@ -0,0 +1,1294 @@
+            *data.shape[:-1], num_pulses, samples_per_pulse)
+
+    def _reshape_flat_pulses(self, data, out, pulse_ids, samples_per_pulse,
+                             first_pulse_offset=None):
+        """Reshape traces by train into traces by pulses."""
+
+        return _reshape_flat_pulses(
+            data, out, pulse_ids, samples_per_pulse,
+            first_pulse_offset or self._first_pulse_offset, self._clock_ratio)
+
+    def _build_train_pulse_coords(self, out, pulse_dim, pulse_layout):
+        """Build xarray coordinates for pulses by train."""
+
+        coords = {'trainId': self._raw_key.train_id_coordinates()}
+
+        # TODO: Create this from pulse index
+        if pulse_dim == 'pulseId':
+            coords['pulseId'] = self._pulses.peek_pulse_ids()
+        elif pulse_dim == 'pulseIndex':
+            coords['pulseIndex'] = np.arange(out.shape[1], dtype=np.int32)
+        elif pulse_dim == 'pulseTime':
+            coords['pulseTime'] = np.arange(
+                out.shape[1], dtype=np.int32) * pulse_layout['length'].max()
+        else:
+            raise ValueError(f'invalid pulse dimension `{pulse_dim}`')
+
+        return coords
+
+    def _build_sample_coords(self, data, sample_dim):
+        """Build xarray coordinates for samples."""
+
+        samples = np.arange(data.shape[-1], dtype=np.int32)
+
+        if sample_dim == 'time':
+            return {'time': samples * self.sampling_period}
+        elif sample_dim == 'sample':
+            return {'sample': samples}
+        else:
+            raise ValueError('sample_dim must be `time` or `sample`')
+
+    def _shape_edge_result(self, edges, amplitudes, index_or_coords,
+                           shape, squeeze_edges, with_amplitudes):
+        """Shape output data of edge finding methods."""
+
+        if shape == 'edges':
+            from . import DelayLineDetector
+
+            edges = DelayLineDetector._build_reduced_pd(
+                None, edges, index_or_coords, entry_level='edgeIndex')
+
+            if not with_amplitudes:
+                return edges
+
+            amplitudes = DelayLineDetector._build_reduced_pd(
+                None, amplitudes, index_or_coords, entry_level='edgeIndex')
+
+            return pd.DataFrame({'edge': edges, 'amplitude': amplitudes})
+
+        else:  # trains, pulses, input
+            if squeeze_edges:
+                edges = self._minimize_sparse_array(edges)
+
+                if with_amplitudes:
+                    amplitudes = self._minimize_sparse_array(amplitudes)
+
+            import xarray as xr
+            index_or_coords['edge'] = np.arange(edges.shape[-1])
+
+            edges = xr.DataArray(edges, coords=index_or_coords)
+
+            if not with_amplitudes:
+                return edges
+
+            return xr.Dataset(dict(
+                edges=edges,
+                amplitudes=xr.DataArray(amplitudes, coords=index_or_coords)))
+
+    def _iter_h5_chunks(self):
```

Comment:
This is copied here as it enabled significant performance improvements for `train_raw()` and `pulse_raw`.